src/Processors/Merges/Algorithms/MergingSortedAlgorithm.cpp

```diff
@@ -87,133 +82,131 @@ void MergingSortedAlgorithm::consume(Input & input, size_t source_num)
     cursors[source_num].reset(current_inputs[source_num].chunk.getColumns(), header);
 
-    queue_variants.callOnVariant([&](auto & queue)
+    queue_variants.callOnBatchVariant([&](auto & queue)
     {
         queue.push(cursors[source_num]);
     });
 }
 
 IMergingAlgorithm::Status MergingSortedAlgorithm::merge()
 {
-    IMergingAlgorithm::Status result = queue_variants.callOnVariant([&](auto & queue)
+    IMergingAlgorithm::Status result = queue_variants.callOnBatchVariant([&](auto & queue)
     {
-        return mergeImpl(queue);
+        return mergeBatchImpl(queue);
     });
 
     return result;
 }
 
-template <typename TSortingHeap>
-IMergingAlgorithm::Status MergingSortedAlgorithm::mergeImpl(TSortingHeap & queue)
+
+template <typename TSortingQueue>
+IMergingAlgorithm::Status MergingSortedAlgorithm::mergeBatchImpl(TSortingQueue & queue)
 {
     /// Take rows in required order and put them into `merged_data`, while the rows are no more than `max_block_size`
     while (queue.isValid())
     {
         if (merged_data.hasEnoughRows())
             return Status(merged_data.pull());
 
-        auto current = queue.current();
+        auto [current_ptr, initial_batch_size] = queue.current();
+        auto current = *current_ptr;
 
-        if (current.impl->isLast() && current_inputs[current.impl->order].skip_last_row)
+        bool batch_skip_last_row = false;
+        if (current.impl->isLast(initial_batch_size) && current_inputs[current.impl->order].skip_last_row)
         {
-            /// Get the next block from the corresponding source, if there is one.
-            queue.removeTop();
-            return Status(current.impl->order);
+            batch_skip_last_row = true;
+
+            if (initial_batch_size == 1)
+            {
+                /// Get the next block from the corresponding source, if there is one.
+                queue.removeTop();
+                return Status(current.impl->order);
+            }
+        }
+
+        UInt64 merged_rows = merged_data.mergedRows();
+        size_t updated_batch_size = initial_batch_size;
+
+        if (merged_rows + updated_batch_size > merged_data.maxBlockSize())
+        {
+            batch_skip_last_row = false;
+            updated_batch_size -= merged_rows + updated_batch_size - merged_data.maxBlockSize();
         }
 
-        /** And what if the block is totally less or equal than the rest for the current cursor?
-            * Or is there only one data source left in the queue? Then you can take the entire block on current cursor.
-            */
-        if (current.impl->isFirst()
-            && !current_inputs[current.impl->order].skip_last_row /// Ignore optimization if last row should be skipped.
-            && (queue.size() == 1
-                || (queue.size() >= 2 && current.totallyLessOrEquals(queue.nextChild()))))
+        bool limit_reached = false;
+        if (limit && merged_rows + updated_batch_size > limit)
```

Comment:
I guess it's just a little bit easier to read
```suggestion
        if (limit && ((merged_rows + updated_batch_size) > limit))
```